core/keystore/src/lib.rs

```diff
@@ -14,185 +14,180 @@
+	/// Adds a session key pair from a seed. Used for testing.
+	pub fn set_key_from_seed<TPair: Pair>(&self, seed: &str) -> Result<()> {
+		let pair = TPair::from_string(seed, None)
+			.map_err(|_| Error::InvalidSeed)?;
+		self.set_key(&pair);
+		Ok(())
 	}
 
-	/// Generate a new key, placing it into the store.
-	pub fn generate<TPair: Pair>(&self, password: &str) -> Result<TPair> {
-		let (pair, phrase, _) = TPair::generate_with_phrase(Some(password));
-		let mut file = File::create(self.key_file_path::<TPair>(&pair.public()))?;
-		::serde_json::to_writer(&file, &phrase)?;
-		file.flush()?;
-		Ok(pair)
+	/// Creates a local session store.
+	pub fn create_local_store<TPair>(&self) -> LocalStore<TPair>
+	where
+		TPair: Pair + Send + Clone,
+		TPair::Public: Send + Clone,
+	{
+		let session_store = LocalStore::new();
+		self.handlers.lock().push((TPair::KEY_TYPE, Box::new(session_store.clone())));
+		session_store
 	}
+}
 
-	/// Create a new key from seed. Do not place it into the store.
-	pub fn generate_from_seed<TPair: Pair>(&mut self, seed: &str) -> Result<TPair> {
-		let pair = TPair::from_string(seed, None)
-			.ok().ok_or(Error::InvalidSeed)?;
-		self.insert_pair(&pair);
-		Ok(pair)
-	}
+/// A local session store supporting a single crypto algorithm.
+#[derive(Clone)]
+pub struct LocalStore<TPair: Pair> {
+	keys: Arc<Mutex<HashMap<TPair::Public, (TPair, Instant)>>>,
+}
 
-	/// Load a key file with given public key.
-	pub fn load<TPair: Pair>(&self, public: &TPair::Public, password: &str) -> Result<TPair> {
-		if let Some(pair) = self.get_pair(public)? {
-			return Ok(pair)
-		}
+impl<TPair> LocalStore<TPair>
+where
+	TPair: Pair + Send + Clone,
+	TPair::Public: Send + Clone,
+{
+	/// Creates a new `LocalStore`.
+	pub fn new() -> Self {
+		let session_store = Self {
+			keys: Default::default(),
+		};
+		session_store
+	}
 
-		let path = self.key_file_path::<TPair>(public);
-		let file = File::open(path)?;
+	/// Gets the key matching the public key.
+	pub fn get_key(&self, public: &TPair::Public) -> Option<TPair> {
+		let pair = {
+			let mut keys = self.keys.lock();
+			if let Some(value) = keys.get_mut(public) {
+				value.1 = Instant::now();
+				Some(value.0.clone())
+			} else {
+				None
+			}
+		};
+		self.clean();
+		pair
+	}
 
-		let phrase: String = ::serde_json::from_reader(&file)?;
-		let (pair, _) = TPair::from_phrase(&phrase, Some(password))
-			.ok().ok_or(Error::InvalidPhrase)?;
-		if &pair.public() != public {
-			return Err(Error::InvalidPassword);
-		}
-		Ok(pair)
+	/// Find keys matching a predicate.
+	pub fn get_keys(&self, f: impl Fn(&TPair::Public) -> bool) -> Vec<TPair> {
+		let filtered_keys = {
+			let mut keys = self.keys.lock();
+			let filtered_keys = keys.iter()
+				.filter(|(public, _)| f(public))
+				.map(|(_, (pair, _))| pair.clone())
+				.collect::<Vec<_>>();
+			for key in &filtered_keys {
+				keys.get_mut(&key.public())
+					.expect("We hold a lock; qed").1 = Instant::now();
```

Comment:
Well the proof should rather be:
```suggestion
					.expect("All filtered_keys are known to exist in `keys`; qed").1 = Instant::now();
```